lib/net.c

```diff
@@ -95,73 +95,87 @@ randomize_rss_key(int guarantee_random_entropy)
 uint8_t rss_key_be[RTE_DIM(default_rss_key)];
 
 /*
- * @ether_type should be passed in host ordering, but is converted
- * to little endian ordering before being added as a filter. The
- * EType register's value should be in little endian, according to
- * the 82599 datasheet.
+ * Add a filter that steers packets to queues based on their EtherType.
+ *
+ * Tecnically, the DPDK rte_flow API allows filters to be specified
+ * on any field in an Ethernet header, but in practice, drivers implement
+ * the RTE_FLOW_ITEM_TYPE_ETH using the EtherType filters available
+ * in hardware. Typically, EtherType filters only support destination
+ * MAC addresses and the EtherType field. We choose to only allow
+ * the EtherType field to be specified, since the destination MAC
+ * address may be extraneous anyway (#74).
  *
- * Does the endianness change with other NICs?
- * 	We have checked the source code of three DPDK networking drivers:
- *	i40e, ixgbe, e1000. And all of them use little endian order.
- *	However, we didn't find the parts for other drivers. We tried to
- *	ask help from the DPDK mailinglist, but didn't get reply.
+ * @ether_type should be passed in host ordering, but is converted
+ * to big endian ordering before being added as a filter, as
+ * required by the rte_flow API. Individual device drivers can then
+ * convert it to whatever endianness is needed.
  */
 int
-ethertype_filter_add(struct gatekeeper_if *iface, uint16_t ether_type,
+ethertype_flow_add(struct gatekeeper_if *iface, uint16_t ether_type,
 	uint16_t queue_id)
 {
-	struct rte_eth_ethertype_filter filter = {
-		.ether_type = rte_cpu_to_le_16(ether_type),
-		.flags = 0,
-		.queue = queue_id,
+	struct rte_flow_attr attr = { .ingress = 1 };
+	struct rte_flow_action_queue queue = { .index = queue_id };
+	struct rte_flow_action action[2] = {
+		{
+			.type = RTE_FLOW_ACTION_TYPE_QUEUE,
+			.conf = &queue,
+	       	},
+		{
+			.type = RTE_FLOW_ACTION_TYPE_END,
+		}
 	};
-	unsigned int i;
+	struct rte_flow_item pattern[2] = {
+		{
+			.type = RTE_FLOW_ITEM_TYPE_ETH,
+			.spec = {
+				.type = rte_cpu_to_be_16(ether_type),
+			},
+			.mask = {
+				.type = 0xFFFF,
+			},
+		},
+		{
+			.type = RTE_FLOW_ITEM_TYPE_END,
+		},
+	};
+	struct rte_flow *flows[iface->num_ports];
+	struct rte_flow_item_eth eth_spec;
+	struct rte_flow_item_eth eth_mask;
```

Comment:
Variables `eth_spec` and `eth_mask` must be declared and initialized before variable `pattern`. Then `pattern[0]` must refer to them since `.spec` and `.mask` are pointers, not structs.

The same must be done with `pattern[0]` in `ipv4_flow_add()`.